<span style="color: #60B5FC; font-weight: bold; font-size: 24px;">Homework 20231002</span>

<span style="color: navy; font-weight: bold; font-size: 20px;">Libraries:</span>

In [63]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import mutual_info_score
from sklearn.feature_extraction import DictVectorizer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer


In [64]:
df = pd.read_csv('/Users/anagilabert/repos/machine-learning-zoomcamp/homework_ana/data/raw.githubusercontent.com_alexeygrigorev_mlbookcamp-code_master_chapter-02-car-price_data.csv')
print(f'There are {len(df)} rows')
display(df.head(2))

There are 11914 rows


Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650

In [65]:
df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [66]:
# Replace NaN in 'Engine HP' and 'Engine Cylinders' by 0
df['Engine HP'].fillna(0, inplace=True)
df['Engine Cylinders'].fillna(0, inplace=True)

target_column = "above_average"

print(df.head())

  Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         C

target: MSRP

<span style="color: darkred; font-weight: bold; font-size: 18px;">Features:</span>

For the rest of the homework, you'll need to use only these columns:

Make,
Model,
Year,
Engine HP,
Engine Cylinders,
Transmission Type,
Vehicle Style,
highway MPG,
city mpg,
MSRP

In [67]:
selected_columns = ['Make', 'Model', 'Year', 'Engine HP', 'Engine Cylinders', 'Transmission Type', 'Vehicle Style', 'highway MPG', 'city mpg', 'MSRP']

In [68]:
df[selected_columns]['MSRP'].isna().sum()

0

In [69]:
df[selected_columns]['MSRP'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 11914 entries, 0 to 11913
Series name: MSRP
Non-Null Count  Dtype
--------------  -----
11914 non-null  int64
dtypes: int64(1)
memory usage: 93.2 KB


In [70]:
# Define a function to train and evaluate a linear regression model
def train_and_evaluate(df_train, df_val, target_column):
    X_train = df_train.drop(columns=[target_column])  # We delete target column 
    y_train = df_train[target_column]  # Target variable
    X_val = df_val.drop(columns=[target_column])
    y_val = df_val[target_column]

    model = LinearRegression()
    model.fit(X_train, y_train)

    y_pred = model.predict(X_val)
    rmse = sqrt(mean_squared_error(y_val, y_pred))
    return rmse

In [71]:
df = df[selected_columns]
df.head(2)

Make       Model  Year  Engine HP  Engine Cylinders Transmission Type  \
0  BMW  1 Series M  2011      335.0               6.0            MANUAL   
1  BMW    1 Series  2011      300.0               6.0            MANUAL   

  Vehicle Style  highway MPG  city mpg   MSRP  
0         Coupe           26        19  46135  
1   Convertible           28        19  40650

<span style="color: #AC1555; font-weight: bold; font-size: 18px;">Question 1:</span>

What is the most frequent observation (mode) for the column transmission_type?

In [72]:
transmission_counts = df['Transmission Type'].value_counts()
print(transmission_counts)

Transmission Type
AUTOMATIC           8266
MANUAL              2935
AUTOMATED_MANUAL     626
DIRECT_DRIVE          68
UNKNOWN               19
Name: count, dtype: int64


The most frecuent is AUTOMATIC

<span style="color: #AC1555; font-weight: bold; font-size: 18px;">Question 2:</span>

What are the two features that have the biggest correlation in this dataset?


In [73]:
variable_types = df.dtypes
variable_types

Make                  object
Model                 object
Year                   int64
Engine HP            float64
Engine Cylinders     float64
Transmission Type     object
Vehicle Style         object
highway MPG            int64
city mpg               int64
MSRP                   int64
dtype: object

In [74]:
cat_columns = df.select_dtypes(include=['object'])


In [75]:
num_columns = df.select_dtypes(include=['int64', 'float64'])
corr_matrix = num_columns.corr()
corr_matrix.style.background_gradient().highlight_max(color = 'darkred')


In [76]:
corr_matrix.unstack().sort_values(ascending = False)[:15]


Year              Year                1.000000
city mpg          city mpg            1.000000
Engine Cylinders  Engine Cylinders    1.000000
Engine HP         Engine HP           1.000000
MSRP              MSRP                1.000000
highway MPG       highway MPG         1.000000
city mpg          highway MPG         0.886829
highway MPG       city mpg            0.886829
Engine HP         Engine Cylinders    0.774851
Engine Cylinders  Engine HP           0.774851
MSRP              Engine HP           0.650095
Engine HP         MSRP                0.650095
Engine Cylinders  MSRP                0.526274
MSRP              Engine Cylinders    0.526274
Year              Engine HP           0.338714
dtype: float64

highway MPG y city mpg : They have a correlation of 0.8868. This means that as highway MPG increases, city mpg also tends to increase.

<span style="color: darkred; font-weight: bold; font-size: 18px;">Make price binary</span>

Now we need to turn the price variable from numeric into a binary format.

- Let's create a variable above_average which is 1 if the price is above its mean value and 0 otherwise.
- Split the data
- Split your data in train/val/test sets with 60%/20%/20% distribution.
- Use Scikit-Learn for that (the train_test_split function) and set the seed to 42.
- Make sure that the target value (above_average) is not in your dataframe.


In [77]:
def split_train_test(X,y):

    # Split the data into a training set (60%)
    X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42) #random_state=seed

    # Split the remaining data into a validation set (20%) and a test set (20%)
    X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42) #0.5 is half training and half test

    df_train = pd.concat([X_train, y_train], axis=1)
    df_val = pd.concat([X_val, y_val], axis=1)
    df_test = pd.concat([X_test, y_test], axis=1)

    return X_train, y_train, X_test, y_test, X_val, y_val, df_train, df_test, df_val

In [78]:
X = df.drop(columns=['MSRP'])
y = df['MSRP']

In [79]:
# PROCESSING DATASET:
X_train, y_train, X_test, y_test, X_val, y_val, df_train, df_test, df_val = split_train_test(X,y)

msrp_mean = df_train['MSRP'].mean()

df[target_column] = np.where(df['MSRP']>=msrp_mean,1,0)

X = df.drop(columns=['MSRP', target_column])
y = df[target_column]



In [80]:
X_train, y_train, X_test, y_test, X_val, y_val, df_train, df_test, df_val = split_train_test(X,y)



In [81]:
df_train.head()

Make            Model  Year  Engine HP  Engine Cylinders  \
4949      Nissan         Frontier  2015      261.0               6.0   
579         FIAT             500L  2016      160.0               4.0   
8364        Ford           Ranger  2011      207.0               6.0   
8775   Chevrolet             S-10  2003      120.0               4.0   
7370  Mitsubishi  Outlander Sport  2016      168.0               4.0   

     Transmission Type        Vehicle Style  highway MPG  city mpg  \
4949         AUTOMATIC      Crew Cab Pickup           21        15   
579             MANUAL                Wagon           33        25   
8364            MANUAL  Extended Cab Pickup           21        16   
8775            MANUAL  Extended Cab Pickup           25        19   
7370         AUTOMATIC              4dr SUV           27        22   

      above_average  
4949              0  
579               0  
8364              0  
8775              0  
7370              0

<span style="color: #AC1555; font-weight: bold; font-size: 18px;">Question 3:</span>

- Calculate the mutual information score between above_average and other categorical variables in our dataset. 
- Use the training set only.
- Round the scores to 2 decimals using round(score, 2).

Which of these variables has the lowest mutual information score?

In [82]:
# DEFINING FUNCTION:
def calculate_mi(feature_series, target_series):
    return mutual_info_score(feature_series, target_series)

# IMPLEMENTATION:
num_columns_cat = df_train.select_dtypes(include=['object'])
target_series = df_train[target_column]

mi_scores = {}

# Calcular la información mutua para cada variable categórica
for column in num_columns_cat:
    feature_series = df_train[column]
    mi_score = calculate_mi(feature_series, target_series)
    mi_scores[column] = mi_score


# Ordenar el diccionario por valores en orden ascendente
sorted_mi_scores = sorted(mi_scores.items(), key=lambda x: x[1])

# Imprimir los resultados ordenados de menor a mayor
for column, score in sorted_mi_scores:
    print(f"Mutual Information between '{column}' and {target_column}: {round(score, 4)}")

Mutual Information between 'Transmission Type' and above_average: 0.0212
Mutual Information between 'Vehicle Style' and above_average: 0.083
Mutual Information between 'Make' and above_average: 0.2354
Mutual Information between 'Model' and above_average: 0.4639


The lowest is 'Transmission Type'.

<span style="color: #AC1555; font-weight: bold; font-size: 18px;">Question 4:</span>

Now let's train a logistic regression.

- Remember that we have several categorical variables in the dataset. Include them using one-hot encoding.
- Fit the model on the training dataset.
- To make sure the results are reproducible across different versions of Scikit-Learn, fit the model with these parameters:

  model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)

Calculate the accuracy on the validation dataset and round it to 2 decimal digits.

In [83]:
X_train_tmp = X_train.copy()
X_val_tmp = X_val.copy()

# Convert the DataFrames to dictionaries
X_train_dict = X_train_tmp.to_dict(orient='records')
X_val_dict = X_val_tmp.to_dict(orient='records')

# Initialize DictVectorizer
dv = DictVectorizer(sparse=False)

# Fit and transform X_train_dict using DictVectorizer
X_train_encoded = dv.fit_transform(X_train_dict)

# Transform X_val_dict using the same DictVectorizer instance
X_val_encoded = dv.transform(X_val_dict)

# Create a logistic regression model
model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)

# Fit the model
model.fit(X_train_encoded, y_train)

# Predict and calculate accuracy on the validation set
y_pred = model.predict(X_val_encoded)
accuracy = round(accuracy_score(y_val, y_pred), 5)

accuracy

0.93957

<span style="color: #AC1555; font-weight: bold; font-size: 18px;">Question 5:</span>

Let's find the least useful feature using the feature elimination technique.

- Train a model with all these features (using the same parameters as in Q4).
- Now exclude each feature from this set and train a model without it. Record the accuracy for each model.
- For each feature, calculate the difference between the original accuracy and the accuracy without the feature.

Which of following feature has the smallest difference?

In [84]:
from sklearn.feature_extraction import DictVectorizer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize a dictionary to store accuracy differences
accuracy_differences = {}

for feature in X_train.columns.tolist():
    # Drop the current feature from both train and validation sets
    X_train_tmp = X_train.drop([feature], axis=1)
    X_val_tmp = X_val.drop([feature], axis=1)

    # Convert the DataFrames to dictionaries
    X_train_dict = X_train_tmp.to_dict(orient='records')
    X_val_dict = X_val_tmp.to_dict(orient='records')

    # Initialize DictVectorizer
    dv = DictVectorizer(sparse=False)

    # Fit and transform X_train_dict using DictVectorizer
    X_train_encoded = dv.fit_transform(X_train_dict)

    # Transform X_val_dict using the same DictVectorizer instance
    X_val_encoded = dv.transform(X_val_dict)

    # Create a logistic regression model
    model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)

    # Fit the model
    model.fit(X_train_encoded, y_train)

    # Predict and calculate accuracy on the validation set
    y_pred = model.predict(X_val_encoded)
    accuracy = round(accuracy_score(y_val, y_pred), 5)

    # Calculate the difference
    accuracy_difference = round(accuracy_global - accuracy, 5)

    # Store the accuracy difference in the dictionary
    accuracy_differences[feature] = accuracy_difference

# Print the accuracy differences
for feature, difference in accuracy_differences.items():
    print(f'Accuracy difference without feature {feature}: {difference}')


Accuracy difference without feature Make: -0.00293
Accuracy difference without feature Model: 0.02561
Accuracy difference without feature Year: -0.00545
Accuracy difference without feature Engine HP: 0.01008
Accuracy difference without feature Engine Cylinders: 0.00295
Accuracy difference without feature Transmission Type: -0.00629
Accuracy difference without feature Vehicle Style: 0.00295
Accuracy difference without feature highway MPG: 0.00211
Accuracy difference without feature city mpg: -0.00041


<span style="color: #AC1555; font-weight: bold; font-size: 18px;">Question 6:</span>

- For this question, we'll see how to use a linear regression model from Scikit-Learn.
- We'll need to use the original column price. Apply the logarithmic transformation to this column.
- Fit the Ridge regression model on the training data with a solver 'sag'. Set the seed to 42.
- This model also has a parameter alpha. Let's try the following values: [0, 0.01, 0.1, 1, 10].
- Round your RMSE scores to 3 decimal digits.
  
Which of these alphas leads to the best RMSE on the validation set?

In [87]:
import numpy as np
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error
from math import sqrt
import pandas as pd

# Define the alpha values to try
alphas = [0, 0.01, 0.1, 1, 10]

# Initialize variables to keep track of the best alpha and RMSE
best_alpha = None
best_rmse = float('inf')

# Apply logarithmic transformation to the target column
df_train['log_target_column'] = np.log1p(df_train[target_column])
df_val['log_target_column'] = np.log1p(df_val[target_column])

# Preprocess categorical columns with one-hot encoding
categorical_columns = X_train.select_dtypes(include=['object']).columns.tolist()
X_train = pd.get_dummies(X_train, columns=categorical_columns, drop_first=True)
X_val = pd.get_dummies(X_val, columns=categorical_columns, drop_first=True)

# Reorder the columns in the validation dataset to match the order in the training dataset
X_val = X_val[X_train.columns]

for alpha in alphas:
    # Initialize and fit Ridge regression model with the current alpha
    model = Ridge(alpha=alpha, solver='sag', random_state=42, max_iter=1000)  # Increase max_iter
    model.fit(X_train, y_train)
    
    # Predict on the validation set
    y_val_pred = model.predict(X_val)
    
    # Calculate RMSE for the current alpha
    rmse = sqrt(mean_squared_error(y_val, y_val_pred))
    
    # Check if this alpha leads to a better RMSE
    if rmse < best_rmse:
        best_alpha = alpha
        best_rmse = rmse

# Round the best RMSE to 3 decimal digits
best_rmse = round(best_rmse, 3)

print(f"The best alpha for Ridge regression is {best_alpha}")
print(f"RMSE on the validation set with the best alpha: {best_rmse}")


/Users/anagilabert/repos/machine-learning-zoomcamp/.conda/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/Users/anagilabert/repos/machine-learning-zoomcamp/.conda/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/Users/anagilabert/repos/machine-learning-zoomcamp/.conda/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/Users/anagilabert/repos/machine-learning-zoomcamp/.conda/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


The best alpha for Ridge regression is 0
RMSE on the validation set with the best alpha: 0.306


/Users/anagilabert/repos/machine-learning-zoomcamp/.conda/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
